src/compression.cpp

```diff
@@ -97,33 +98,56 @@ namespace sparrow_ipc
             return decompressed_data;
         }
 
-        // TODO These functions could be moved to serialize_utils and deserialize_utils if preferred
-        // as they are handling the header size
-        std::vector<std::uint8_t> uncompressed_data_with_header(std::span<const std::uint8_t> data)
+        void insert_compressed_data(std::vector<uint8_t>& result, std::int64_t original_size, const std::vector<uint8_t>& compressed_body)
+        {
+            result.reserve(details::CompressionHeaderSize + compressed_body.size());
+            result.insert(result.end(), reinterpret_cast<const uint8_t*>(&original_size), reinterpret_cast<const uint8_t*>(&original_size) + sizeof(original_size));
+            result.insert(result.end(), compressed_body.begin(), compressed_body.end());
+        }
+
+        void insert_uncompressed_data(std::vector<uint8_t>& result, const std::span<const uint8_t>& data)
         {
-            std::vector<std::uint8_t> result;
-            result.reserve(details::CompressionHeaderSize + data.size());
             const std::int64_t header = -1;
+            result.reserve(details::CompressionHeaderSize + data.size());
             result.insert(result.end(), reinterpret_cast<const uint8_t*>(&header), reinterpret_cast<const uint8_t*>(&header) + sizeof(header));
             result.insert(result.end(), data.begin(), data.end());
-            return result;
         }
 
-        std::vector<std::uint8_t> compress_with_header(std::span<const std::uint8_t> data, compress_func comp_func)
+        std::span<const std::uint8_t> compress_with_header(
+            const std::span<const std::uint8_t>& data,
+            compress_func comp_func,
+            compression_cache_t& cache)
         {
+            const void* buffer_ptr = data.data();
```

Comment:
Hmm, what if at some point we pass a different slice starting at the same pointer? It might produce a false positive, perhaps we want the map key to be a (pointer, size) tuple?